libraries/AP_OpticalFlow/AP_OpticalFlow_UAVCAN.cpp

```diff
@@ -0,0 +1,110 @@
+#include <AP_HAL/AP_HAL.h>
+
+#if HAL_WITH_UAVCAN
+
+#include <AP_Common/Semaphore.h>
+
+#include "AP_OpticalFlow_UAVCAN.h"
+
+#include <AP_BoardConfig/AP_BoardConfig_CAN.h>
+#include <AP_UAVCAN/AP_UAVCAN.h>
+
+#include <com/hex/equipment/flow/Measurement.hpp>
+
+extern const AP_HAL::HAL& hal;
+
+#define debug_flow_uavcan(level_debug, can_driver, fmt, args...) do { if ((level_debug) <= AP::can().get_debug_level_driver(can_driver)) { hal.console->printf(fmt, ##args); }} while (0)
+
+//UAVCAN Frontend Registry Binder
+UC_REGISTRY_BINDER(MeasurementCb, com::hex::equipment::flow::Measurement);
+
+uint8_t AP_OpticalFlow_UAVCAN::_node_id = 0;
+AP_OpticalFlow_UAVCAN* AP_OpticalFlow_UAVCAN::_driver = nullptr;
+AP_UAVCAN* AP_OpticalFlow_UAVCAN::_ap_uavcan = nullptr;
+/*
+  constructor - registers instance at top Flow driver
+ */
+AP_OpticalFlow_UAVCAN::AP_OpticalFlow_UAVCAN(OpticalFlow &flow) :
+    OpticalFlow_backend(flow)
+{}
+
+void AP_OpticalFlow_UAVCAN::subscribe_msgs(AP_UAVCAN* ap_uavcan)
+{
+    if (ap_uavcan == nullptr) {
+        return;
+    }
+
+    auto* node = ap_uavcan->get_node();
+
+    uavcan::Subscriber<com::hex::equipment::flow::Measurement, MeasurementCb> *measurement_listener;
+    measurement_listener = new uavcan::Subscriber<com::hex::equipment::flow::Measurement, MeasurementCb>(*node);
+    // Msg Handler
+    const int measurement_listener_res = measurement_listener->start(MeasurementCb(ap_uavcan, &handle_measurement));
+    if (measurement_listener_res < 0) {
+        AP_HAL::panic("UAVCAN Flow subscriber start problem\n\r");
+        return;
+    }
+}
+
+AP_OpticalFlow_UAVCAN* AP_OpticalFlow_UAVCAN::detect(OpticalFlow &flow)
+{
+    if (_ap_uavcan == nullptr) {
+        return nullptr;
+    }
+    if (_driver == nullptr) {
+        _driver = new AP_OpticalFlow_UAVCAN(flow);
+        if (_driver == nullptr) {
```

Comment:
If we failed to allocate the memory the first time, why are we likely to the second time? Given our use case this seems like an excessive usage of code.